### PR TITLE
Improvement: Faster topological sort for Time Series Causal Graph

### PR DESCRIPTION
**Pull request by matteo-causalens**
## Description and Motivation
Improved the topological sort for Time Series Causal Graph to avoid to compute all the topological sorts if only one is needed.

To do it, we'll use  https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.dag.lexicographical_topological_sort.html.

We also added a new boolean keyword argument `respect_time_ordering` that defaults at true. If False, then the `CausalGraph`'s topological sort will be used.

## Additional Context
This issue was raise by Max. The solution was discussed with Andrew offline on Slack.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Improvement (non-breaking change which improves functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Refactor (moving code around, general refactoring improvements)
- [ ] Documentation (documentation)

## PR Checklist:

- [ ] I have reviewed my own PR? (review the PR as you are reviewing someone else's PR)
- [ ] I have implemented all requirements? (see JIRA, project documentation)
- [ ] I am affecting someone else's work? If so: I have added those people as reviewers?
- [ ] I have added comments to all the bits that are hard to follow
- [ ] At a bare minimum, I tested this manually
- [ ] I have added unit test(s)
- [ ] Is this a bugfix? If so have I added a regression test?
- [ ] Does this PR satisfy the [one PR, one concern](https://medium.com/@fagnerbrack/one-pull-request-one-concern-e84a27dfe9f1) rule?
- [ ] If needed, documentation has been added/updated?
- [ ] I have updated the appropriate changelog with a line for my changes
- [ ] If this PR breaks reproducibility, have I added a note in the changelog explaining the break in reproducibility
      and its extent?

### CODE REVIEW

File: cai_causal_graph/time_series_causal_graph.py
Comment (matteo-causalens) [inline cai_causal_graph/time_series_causal_graph.py:304]: Placeholder for myself to add the changelog

**Review by tadhgokeeffe — APPROVED** [review]
LGTM

File: cai_causal_graph/time_series_causal_graph.py
Comment (andrew-causalens) [inline cai_causal_graph/time_series_causal_graph.py:333]: I think this always needs to be checked so would do it after `if respect_time_ordering` but before the next `if` block. No need to do in the `else` that calls `super` as it is checked in `super`.

File: tests/test_time_series_graph.py
Comment (andrew-causalens) [inline tests/test_time_series_graph.py:1]: Do you mind adding the graphs I put in the email? Think it is worth just adding a few more unit tests.

**Review by andrew-causalens — APPROVED** [review]
LGTM. Just 2 comments to address but no need to hold up approval.

File: cai_causal_graph/time_series_causal_graph.py
Comment (matteo-causalens) [reply cai_causal_graph/time_series_causal_graph.py:333]: why? when `return_all=True` it will be checked by `super().get_topological_order(return_all=return_all)` if I am not mistaken

File: tests/test_time_series_graph.py
Comment (matteo-causalens) [reply tests/test_time_series_graph.py:1]: of course will do it now

File: cai_causal_graph/time_series_causal_graph.py
Comment (andrew-causalens) [reply cai_causal_graph/time_series_causal_graph.py:333]: Good point. I missed that `super` was called there as well. My bad.

Comment (andrew-causalens) [reply cai_causal_graph/time_series_causal_graph.py:333]: Would just add a comment that this check is not needed in other conditionals as it is done by `super`.

File: cai_causal_graph/time_series_causal_graph.py
Comment (ilya-causalens) [inline cai_causal_graph/time_series_causal_graph.py:313]: missing return description.

File: cai_causal_graph/time_series_causal_graph.py
Comment (ilya-causalens) [inline cai_causal_graph/time_series_causal_graph.py:327]: I don't really understand why you are checking `if tmp not in ordered_nodes`? Why is it not sufficient to just check `if len(tmp) > 0`?

As far as I understand, the `ordered_nodes_list` contains unique topological orders. The `self._get_time_topological_order` will either return the input order unchanged (if it respects time) or return an empty list. Hence, if a non empty list is returned, then it must be unique.

I think checking `if tmp not in ordered_nodes` may become somewhat computationally expensive (relatively speaking) when you have a huge graph with a lot of different topological sorts.

**Review by ilya-causalens — APPROVED** [review]
Just a minor comment on efficiency, but other than that looks good.

File: cai_causal_graph/time_series_causal_graph.py
Comment (matteo-causalens) [reply cai_causal_graph/time_series_causal_graph.py:327]: There could be duplicates. I remember it from the previous implementation. But, for the sake of efficiency we could remove the check and leave duplicates since it's not conceptually wrong

Comment (matteo-causalens) [reply cai_causal_graph/time_series_causal_graph.py:327]: Happy to remove the check

Comment (ilya-causalens) [reply cai_causal_graph/time_series_causal_graph.py:327]: Well thing is if there are duplicates, then you should also be removing them below when `respect_time_order = False`. Otherwise, it isn't consistent?

I don't mind (and have approved the PR), just seemed off to me. I would say - either you should check for duplicates in all cases, or not check for them at all.

Comment (andrew-causalens) [reply cai_causal_graph/time_series_causal_graph.py:338]: I would not return any duplicates. The list of valid top sorts should be unique.

Comment (matteo-causalens) [reply cai_causal_graph/time_series_causal_graph.py]: Sorry, I was confusing the old implementation where I was sorting. Here we filter so no need to check.

File: tests/test_time_series_graph.py
Comment (andrew-causalens) [inline tests/test_time_series_graph.py:1161]: Can you add a test where respect time is False and show it is length of 2 and the other valid one is there?

**Review by andrew-causalens — APPROVED** [review]

**Pull request merged**
